src/ape/pytest/contextmanagers.py

```diff
@@ -1,26 +1,118 @@
+from collections import deque
+import re
 from typing import Optional, Type
 
+import ape
 from ape.exceptions import ContractLogicError, TransactionError
 
 
 class RevertsContextManager:
-    def __init__(self, expected_message: Optional[str] = None):
+    def __init__(self, expected_message: Optional[str] = None, dev_message: Optional[str] = None):
+        if dev_message is not None and expected_message is not None:
+            raise ValueError("Cannot supply both 'expected_message' and 'dev_message'.")
+
         self.expected_message = expected_message
+        self.dev_message = dev_message
 
-    def __enter__(self):
-        pass
+    def _check_dev_message(self, exc: ContractLogicError):
+        """
+        Attempts to extract a dev-message from the contract source code by inspecting what
+        instruction(s) led to a transaction revert.
 
-    def __exit__(self, exc_type: Type, exc_value: Exception, traceback) -> bool:
-        if exc_type is None:
-            raise AssertionError("Transaction did not revert.")
+        TODO: Document exception-raising cases
+        """
+        if exc.txn is None:
+            raise AssertionError("Could not fetch transaction information to check dev message.")
 
-        if not isinstance(exc_value, ContractLogicError):
+        receipt = exc.txn.provider.get_receipt(
+            exc.txn.txn_hash,
+            raise_on_fail=False,
+        )
+
+        try:
+            contract = ape.Contract(receipt.receiver)
+        except ValueError as exc:
             raise AssertionError(
-                f"Transaction did not revert.\n"
-                f"However, an exception of type {type(exc_value)} occurred: {exc_value}"
-            ) from exc_value
+                f"Could not fetch contract at {receipt.receiver} to check dev message"
+            ) from exc
+
+        if contract.contract_type.sourcemap is None:
+            raise AssertionError("Could not fetch contract source to check dev message")
+
+        trace = deque(exc.trace)
```

Comment:
Need to swap this to `list`, originally I was thinking I could just look at the last frame so I was using `deque(exc.trace, maxlen=1)`

Going to check out if maybe we could at least truncate the trace here to something like 10 or so, since it's only ever going to be one Vyper statement that causes a revert. I'll have to play around and see though if there's a suitable `maxlen` or not